File: base/src/com/google/idea/blaze/base/scope/output/IssueOutput.java
/*
 * Copyright 2016 The Bazel Authors. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.idea.blaze.base.scope.output;

import com.google.idea.blaze.base.scope.BlazeContext;
import com.google.idea.blaze.base.scope.Output;
import com.intellij.pom.Navigatable;
import java.io.File;
import javax.annotation.Nullable;

/** An issue in a blaze operation. */
public class IssueOutput implements Output {

  public static final int NO_LINE = -1;
  public static final int NO_COLUMN = -1;

  @Nullable private final File file;
  private final int line;
  private final int column;
  private final Category category;
  private final String message;
  @Nullable Navigatable navigatable;
  @Nullable IssueData issueData;

  /** Base class for issue data */
  public static class IssueData {}

  /** Issue category */
  public enum Category {
    ERROR,
    WARNING,
    STATISTICS,
    INFORMATION
  }

  public static Builder issue(Category category, String message) {
    return new Builder(category, message);
  }

  public static Builder error(String message) {
    return new Builder(Category.ERROR, message);
  }

  public static Builder warn(String message) {
    return new Builder(Category.WARNING, message);
  }

  /** Builder for an issue */
  public static class Builder {
    private final Category category;
    private final String message;
    @Nullable private File file;
    private int line = NO_LINE;
    private int column = NO_COLUMN;
    @Nullable Navigatable navigatable;
    @Nullable IssueData issueData;

    public Builder(Category category, String message) {
      this.category = category;
      this.message = message;
    }

    public Builder inFile(@Nullable File file) {
      this.file = file;
      return this;
    }

    public Builder onLine(int line) {
      this.line = line;
      return this;
    }

    public Builder inColumn(int column) {
      this.column = column;
      return this;
    }

    public Builder withData(@Nullable IssueData issueData) {
      this.issueData = issueData;
      return this;
    }

    public Builder navigatable(@Nullable Navigatable navigatable) {
      this.navigatable = navigatable;
      return this;
    }

    public IssueOutput build() {
      return new IssueOutput(file, line, column, navigatable, category, message, issueData);
    }

    public void submit(BlazeContext context) {
      context.output(build());
      if (category == Category.ERROR) {
        context.setHasError();
      }
    }
  }

  private IssueOutput(
      @Nullable File file,
      int line,
      int column,
      @Nullable Navigatable navigatable,
      Category category,
      String message,
      @Nullable IssueData issueData) {
    this.file = file;
    this.line = line;
    this.column = column;
    this.navigatable = navigatable;
    this.category = category;
    this.message = message;
    this.issueData = issueData;
  }

  @Nullable
  public File getFile() {
    return file;
  }

  public int getLine() {
    return line;
  }

  public int getColumn() {
    return column;
  }

  @Nullable
  public Navigatable getNavigatable() {
    return navigatable;
  }

  public Category getCategory() {
    return category;
  }

  public String getMessage() {
    return message;
  }

  @Override
  public String toString() {
    return message;
  }

  @Nullable
  public IssueData getIssueData() {
    return issueData;
  }
}

File: base/src/com/google/idea/blaze/base/scope/output/IssueOutput.kt
/*
 * Copyright 2016 The Bazel Authors. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
@file:Suppress("UnstableApiUsage")

package com.google.idea.blaze.base.scope.output

import com.google.idea.blaze.base.io.VfsUtils
import com.google.idea.blaze.common.Context
import com.google.idea.blaze.common.Output
import com.intellij.build.events.MessageEvent
import com.intellij.build.issue.BuildIssue
import com.intellij.build.issue.BuildIssueQuickFix
import com.intellij.openapi.fileEditor.OpenFileDescriptor
import com.intellij.openapi.project.Project
import com.intellij.pom.Navigatable
import java.io.File

/** An issue in a blaze operation.  */
class IssueOutput (
  issue: BuildIssue,
  val kind: MessageEvent.Kind,
) : Output, BuildIssue by issue {

  companion object {
    @JvmStatic
    fun issue(kind: MessageEvent.Kind, title: String): Builder {
      return Builder(kind, title)
    }

    @JvmStatic
    fun error(title: String): Builder {
      return Builder(MessageEvent.Kind.ERROR, title)
    }

    @JvmStatic
    fun warn(title: String): Builder {
      return Builder(MessageEvent.Kind.WARNING, title)
    }
  }

  @Deprecated("used only for backwards compatibility")
  fun getMessage(): String = "$title\n$description"

  class Builder(private val kind: MessageEvent.Kind, private val title: String) {
    private var description: String = ""
    private var navigatable: (Project) -> Navigatable? = { null }
    private val fixes: MutableList<Pair<String, BuildIssueQuickFix>> = mutableListOf()

    fun withFile(file: File): Builder = withFile(file, line = 0, column = 0)

    fun withFile(file: File, line: Int = 0, column: Int = 0): Builder {
      val virtualFile = VfsUtils.resolveVirtualFile(file, false)

      if (virtualFile != null) {
        navigatable = { OpenFileDescriptor(it, virtualFile, line, column) }
      }

      return this
    }

    fun withNavigatable(navigatable: Navigatable): Builder {
      this.navigatable = { navigatable }
      return this
    }

    fun withDescription(description: String): Builder {
      this.description = description
      return this
    }

    fun withFix(description: String, fix: BuildIssueQuickFix): Builder {
      fixes.add(description to fix)
      return this
    }

    fun build(): IssueOutput {
      val builder = StringBuilder(description)

      if (fixes.isNotEmpty()) {
        builder.append("\n\nPossible fixes:\n")
      }

      for ((description, fix) in fixes) {
        builder.append("$description (<a href=\"${fix.id}\">click here</a>)")
      }

      val issue = object : BuildIssue {
        override val description: String
          get() = builder.toString()

        override val quickFixes: List<BuildIssueQuickFix>
          get() = fixes.map { it.second }

        override val title: String
          get() = this@Builder.title

        override fun getNavigatable(project: Project): Navigatable? = navigatable(project)
      }

      return IssueOutput(issue, kind)
    }

    fun submit(context: Context<*>) {
      context.output(build())
      if (kind == MessageEvent.Kind.ERROR) {
        context.setHasError()
      }
      if (kind == MessageEvent.Kind.WARNING) {
        context.setHasWarnings()
      }
    }
  }
}
